cpp/src/arrow/flight/sql/server_session_middleware.cc

```diff
@@ -80,7 +80,7 @@ class ServerSessionMiddlewareImpl : public ServerSessionMiddleware {
 
   Status CloseSession() override {
     const std::lock_guard<std::shared_mutex> l(mutex_);
-    if (static_cast<bool>(session_)) {
+    if (!static_cast<bool>(session_)) {
```

Comment:
Yes the C++ server scenario was failing when I added a test to the Go client scenario that attempts to close the currently open session. I'm not especially familiar with C++, but my understanding is that this was previously saying that "if the session DOES exist, then it is an error to close it" which seems backwards. Is my understanding accurate?

Reply:
@indigophox

Reply:
@joellubi is correct here and the fix also checks out.  Looks like from my commit history that I typoed this while hastily updating the code to work around and omit functionality affected by #40071.  This used to be covered in integration testing but was temporarily removed as the aforementioned issue breaks the ability (in C++) to correctly invalidate sessions (on the server side) when the CloseSession Action is called.

Reply:
Notably if you continue on with the same client cookie middleware in any integration test after closing the session from the client, calls are going to get failed out by the C++ server session middleware because the client cookie middleware will submit a now-invalid session token (that the server was unable to invalidate) with the request headers.  This is fixable pending #40071.

Reply:
@joellubi Can you add a corresponding fix to the client middleware that matches #40071 which @indigophox mentioned? Or at minimum, file a follow-up issue for this so we don't forget it

Reply:
@zeroshade I think that in the case @indigophox mentioned the client middleware is actually behaving correctly. The problem with the C++ server middleware is that it's unable to invalidate the cookie for an existing session (i.e. it never sends `Set-Cookie: arrow_flight_session_id=<session-id>; Max-Age=0`) so the client keeps the cookie stored, as it should. Then the next time the client makes any request it sends the cookie (i.e. `Cookie: arrow_flight_session_id=<session-id>`) along with it, but the server has completely forgotten about that session and considers the token invalid.

Reply:
@indigophox As I was writing this, the following idea occurred to me:

What if in the meantime while the C++ issue is getting resolved, we changed the C++ implementation to respond with `CloseSessionResult_NOT_CLOSEABLE` to all `CloseSessionRequest`'s? Admittedly that's not exactly what I imagined that result status would end up being used for, but it's kind of accurate right now and most importantly it gives _some_ information to clients to be able recover the edge case gracefully.

Reply:
@joellubi ultimately this is scoped to the implementation, so anyone building an app against this can do as they please.  The shipped ServerSessionMiddleware just handles the internals; the actual RPC handler can do whatever the app author wants.  I am hoping we can close on #40071 fairly soon (before 16.0.0 code freeze?) so the full functionality can be restored.  I'll resume the discussion over there and hopefully that can get wrapped up with the existing solution or another one.

My other thought for an if-we-have-to-go-down-that-road workaround is for the middleware to internally flag the token as invalidated, so it can expire it next time it's presented instead of outright treating it as invalid, or alternatively (say you're calling SetSessionOptions) just clobber it with a new token for a new session as appropriate.  But again this is dumping work into a temporary hack while we sort out the root issue with middleware handling behaviour.

Reply:
I think this is a good discussion potentially but I agree that this is an issue scoped to the implementation, which doesn't seem to be an issue for us on the Go side currently. I'm going to merge this as it looks pretty well covered and handled to me here but feel free to continue this discussion even with this PR merged if desired.